### PR TITLE
Unset the needs-notify bit in a Notifier when a Future is fetched

### PR DESCRIPTION
**Pull request by TheBlueMatt**
If a `Notifier` gets `notify()`ed and the a `Future` is fetched,
even though the `Future` is marked completed from the start and
the user may pass callbacks which are called, we'll never wipe the
needs-notify bit in the `Notifier`.

This addresses (yet another) Future bug that a user caught :(. For the sake of completeness this also avoids considering a future as complete until we get `poll`'d.

### CODE REVIEW

**Comment by MaxFangX** [comment]
Although I can't comment on the code changes, I have run our failing test using this patch at d61ebec62d019fb7d6a991706d3cfbd3913303f0 and confirmed it resolves the issue we were seeing in our integration tests of `get_persistable_update_future` always returning a `Future` which is `Ready`. 👍 Thanks @TheBlueMatt for looking into this for us, looking forward to merge.

File: lightning/src/util/wakers.rs [hunk 1]
```diff
@@ -151,19 +152,21 @@ impl<F: Fn() + Send> FutureCallback for F {
 }
 
 pub(crate) struct FutureState {
-	callbacks: Vec<Box<dyn FutureCallback>>,
+	// When we're tracking whether a callback counts as having woken the user's code, we check the
+	// first bool - set to false if we're just calling a Waker, and true if we're calilng an actual
```
Comment (dunxen) [inline lightning/src/util/wakers.rs:156]: nit
```suggestion
	// first bool - set to false if we're just calling a Waker, and true if we're calling an actual
```

File: lightning/src/util/wakers.rs [hunk 2]
```diff
@@ -41,45 +55,47 @@ impl Notifier {
 		}
 	}
 
+	fn propagate_future_state_to_notify_flag(&self) -> MutexGuard<(bool, Option<Arc<Mutex<FutureState>>>)> {
+		let mut lock = self.notify_pending.lock().unwrap();
+		if let Some(existing_state) = &lock.1 {
+			if existing_state.lock().unwrap().callbacks_made {
+				// If the existing futurestate has completed and actually made callbacks, consider
```
Comment (dunxen) [inline lightning/src/util/wakers.rs:62]:
```suggestion
				// If the existing future state has completed and actually made callbacks, consider
```

File: lightning/src/util/wakers.rs [hunk 3]
```diff
@@ -409,4 +437,30 @@ mod tests {
 		assert_eq!(Pin::new(&mut future).poll(&mut Context::from_waker(&waker)), Poll::Ready(()));
 		assert_eq!(Pin::new(&mut second_future).poll(&mut Context::from_waker(&second_waker)), Poll::Ready(()));
 	}
+
+	#[test]
+	fn test_dropped_future_doesnt_count() {
+		// Tests that if a Future gets drop'd before it is poll()ed Ready it doesn't count as
```
Comment (dunxen) [inline lightning/src/util/wakers.rs:443]:
```suggestion
		// Tests that if a Future gets drop'd before it is poll()ed `Ready` it doesn't count as
```

File: lightning/src/util/wakers.rs [hunk 4]
```diff
@@ -409,4 +437,30 @@ mod tests {
...
+		// having been woken, leaving the notify-required flag set.
+		let notifier = Notifier::new();
+		notifier.notify();
+
+		// If we get a future and don't touch it we're definitely still notify-required.
+		notifier.get_future();
+		assert!(notifier.wait_timeout(Duration::from_secs(1)));
+		assert!(!notifier.wait_timeout(Duration::from_secs(1)));
+
+		// Even if we poll'd once but didn't onserve a `Ready`, we should be notify-required.
```
Comment (dunxen) [inline lightning/src/util/wakers.rs:453]:
```suggestion
		// Even if we poll'd once but didn't observe a `Ready`, we should be notify-required.
```

**Review by dunxen — COMMENTED** [review]
Approach ACK, just some nits.

**Review by dunxen — APPROVED** [review]

**Pull request merged**